app/src/behaviors/behavior_smart_interrupt.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_smart_interrupt
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/keymap.h>
+#include <zmk/matrix.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+#include <zmk/events/layer_state_changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS 10
+
+struct behavior_smart_interrupt_config {
+    int32_t shared_key_positions_len;
+    int32_t shared_layers_len;
+    struct zmk_behavior_binding *behaviors;
+    int32_t shared_layers[32];
+    int32_t timeout_ms;
+    int32_t shared_key_positions[];
+};
+
+struct active_smart_interrupt {
+    bool is_active;
+    bool is_pressed;
+    bool first_press;
+    uint32_t position;
+    const struct behavior_smart_interrupt_config *config;
+    struct k_work_delayable release_timer;
+    int64_t release_at;
+    bool timer_started;
+    bool timer_cancelled;
+};
+
+static int stop_timer(struct active_smart_interrupt *smart_interrupt) {
+    int timer_cancel_result = k_work_cancel_delayable(&smart_interrupt->release_timer);
+    if (timer_cancel_result == -EINPROGRESS) {
+        // too late to cancel, we'll let the timer handler clear up.
+        smart_interrupt->timer_cancelled = true;
+    }
+    return timer_cancel_result;
+}
+
+static void reset_timer(int32_t timestamp, struct active_smart_interrupt *smart_interrupt) {
+    smart_interrupt->release_at = timestamp + smart_interrupt->config->timeout_ms;
+    int32_t ms_left = smart_interrupt->release_at - k_uptime_get();
+    if (ms_left > 0) {
+        k_work_schedule(&smart_interrupt->release_timer, K_MSEC(ms_left));
+        LOG_DBG("Successfully reset smart interrupt timer");
+    }
+}
+
+void behavior_smart_interrupt_timer_handler(struct k_work *item) {
+    struct active_smart_interrupt *smart_interrupt =
+        CONTAINER_OF(item, struct active_smart_interrupt, release_timer);
+    if (!smart_interrupt->is_active) {
+        return;
+    }
+    if (smart_interrupt->timer_cancelled) {
+        return;
+    }
+    if (smart_interrupt->is_pressed) {
+        return;
+    }
+    LOG_DBG("Smart interrupted deactivated due to timer");
+    smart_interrupt->is_active = false;
+    struct zmk_behavior_binding_event event = {.position = smart_interrupt->position,
+                                               .timestamp = k_uptime_get()};
+    behavior_keymap_binding_pressed(&smart_interrupt->config->behaviors[2], event);
+    behavior_keymap_binding_released(&smart_interrupt->config->behaviors[2], event);
```

Comment:
Do we need/care to have the time between press and release be configurable? Some hosts might not like the "immediate" press and release, for certain uses.

And if that needs to have an optional delay, probably best to use the behavior queue added for combos to queue up the press and release.